Adds ability to remove activity.

Neathan opened the pull request:
**Comment:**
Adds the ability to remove activities. The plan was to also be able to modify the activity but then the issue of activities not being a single object made things harder.

**Notes:**
The remove function added to storage removes every instance for the given day that matches the entry. This should not be a problem but the instance could occur where the exact same activity is registered on the same day and would be removed if either of them would be removed. This could be solved by requiring that the first index of the collection of activities is specified to the remove function and thus limiting the search enough to determine what collection should be removed.

When it comes the ability to modify an activity this would requires a similar solution as suggested above. However, in my opinion a more robust solution would be to change how we save activities. If activities contained the start and end time of each activity that would be easier to manage, data wise. This is however a much larger issue and should be handled on its own.

EMattfolk commented on src/storage/activities.tsx:
```diff
@@ -25,6 +25,34 @@ export const activityDayGt = (a: ActivitiesDay, b: ActivitiesDay): boolean => {
+ */
+const activityDayEq = (a: ActivitiesDay, b: ActivitiesDay): boolean => {
+  return a.date === b.date
+      && a.entries === b.entries
```
This would probably require activityEntryEq to be above.
```suggestion
      && a.entries.length === b.entries.length
      && a.entries.every((e, i) => activityEntryEq(e, b.entries[i]))
```

Neathan replied:
Great suggestion! Applied it in the coming commits.

EMattfolk commented on src/storage/activities.tsx:
```diff
@@ -160,9 +188,33 @@ export const useActivities = (): [Activities, ModifyActivities] => {
+   * @returns A new activity list with the entry removed
+   */
+  const remove = (activity: ActivitiesEntry, day: ActivitiesDay, activities: Activities): Activities => {
+    const dayIndex = activities.findIndex(elem => elem === day);
```
This should use activityDayEq

Neathan replied:
Why was this marked as resolved? @malin-mallan

EMattfolk commented on src/storage/activities.tsx:
```diff
@@ -160,9 +188,33 @@ export const useActivities = (): [Activities, ModifyActivities] => {
+   */
+  const remove = (activity: ActivitiesEntry, day: ActivitiesDay, activities: Activities): Activities => {
+    const dayIndex = activities.findIndex(elem => elem === day);
+    if(dayIndex !== -1) {
```
```suggestion
    if (dayIndex !== -1) {
```

EMattfolk commented on src/storage/activities.tsx:
```diff
@@ -160,9 +188,33 @@ export const useActivities = (): [Activities, ModifyActivities] => {
   const modifyActivities: ModifyActivities = {
     add: add,
     addInterval: addInterval,
+    remove: (activity, day) => setStoreItem(activitiesKey, remove(activity, day, activities)),
```
Unless remove is used somewhere else setStoreItem should be moved inside the function similar to the others.

Neathan replied:
Got it, i was wondering about this when i wrote it. Since my knowledge about the inner workings of the storage module is limited i mainly took inspiration from how the calendar functions worked, thus its similar look. But of course this could be moved in to a separate function to match the current files structure. Will do.

EMattfolk commented on src/components/activity/ActivityRegistrator.tsx:
```diff
@@ -10,36 +10,101 @@ import { ChoiceBasedTextInput } from '../ChoiceBasedTextInput';
 import { ISODate } from 'utils';
+import { getDaysInMonth } from 'react-native-paper-dates/lib/typescript/src/Date/dateUtils';
 
 export const ActivityRegistrator = ({ route, navigation }: any) => {
```
I do feel like we should add typing to these components...

Neathan replied:
I agree, might do add it to this PR if i have the time. Otherwise i'll add an issue for it.

EMattfolk commented on src/components/activity/ActivityHistory.tsx:
```diff
@@ -132,6 +144,7 @@ export const ActivityHistory = ({route, navigation}: any) => {
         contentContainerStyle={{
           paddingBottom: 75
         }}
+        
```
```suggestion
```

EMattfolk commented on src/components/activity/ActivityRegistrator.tsx:
```diff
@@ -10,36 +10,101 @@ import { ChoiceBasedTextInput } from '../ChoiceBasedTextInput';
-  const [values, modifyValues] = Storage.useValues();
-  const [activities, modifyActivities] = Storage.useActivities();
+  // Overwrite default values from entry (if it exists)
+  if (route.params?.entry && route.params?.day) {
```
If route.params.entry is a index in the list much of this logic could be simplified. That would require the history view to be changed slightly, but not much (see that comment).

Neathan replied:
I will look into this. I did not really think about this since my intentions at first was only to add the missing functionality. I took a look at calendar and saw how it was done there, will try to adapt a similar structure here.

EMattfolk commented on src/components/activity/ActivityHistory.tsx:
```diff
@@ -81,13 +89,15 @@ export const ActivityHistory = ({route, navigation}: any) => {
-              <List.Icon icon={activity.icon}/>
-            </View>
-          }/>
+        <Surface key={'hi_' + fromDate.toString()} style={{ borderRadius: 5, elevation: elevation.medium, marginHorizontal: 10 , marginVertical: 5 }}>
+          <TouchableRipple onPress={() => onModifyDate(day, activity)} style={{ flexDirection: 'row' }}>
+            <List.Item style={{flex: 1, flexGrow: 1 }} title={fromTimeString + ' - ' + toTimeString} description={activity.text} right={() =>
+              <View style={{flexDirection: 'row', alignItems: 'center'}}>
+                <Caption>{activity.importance + ' | ' + activity.enjoyment}</Caption>
+                <List.Icon icon={activity.icon}/>
+              </View>
+            }/>
+          </TouchableRipple>
```
Instead of pushing a whole element, only the title and the index of an entry should be pushed. With that information the flatlist can render the components and some of the logic in ActivityRegistrator can be simplified.

EMattfolk commented on src/components/activity/ActivityRegistrator.tsx:
```diff
@@ -10,36 +10,101 @@ import { ChoiceBasedTextInput } from '../ChoiceBasedTextInput';
+    for(let i = 0; i < day.entries.length; ++i) {
+      const currentEntry = day.entries[i];
+      // If element matches entry
+      if(currentEntry === entry) {
```
Entries cannot be compared with the equality operator.

Neathan replied:
This code has been entirely reworked.

EMattfolk reviewed (changes requested):
It looks good for the most part, but I do think it is unacceptable that all of the same entry will be removed. I added some comments on how it could be fixed. Implementing proper typing would also be a plus.

malin-mallan reviewed (changes requested):
I understand how you think but as a user it is weird to come to a editpage and not be able to save the changes. If the user only should be able to remove the activity a dialog like values is a much better solution.

Neathan commented:
> It looks good for the most part, but I do think it is unacceptable that all of the same entry will be removed. I added some comments on how it could be fixed. Implementing proper typing would also be a plus.

I agree that the functionality is not completed, this was meant to be a draft. My intention was to open this PR for input on how this problem would be managed. Since this feature was functionally simple at the surface but required rather substantial storage implementations quite a few new quirks came up with how we currently have structured the data storage. I'm currently working on adding the missing functionality and i think i've come up with a solution to how activities should be modified. This will however once again increase the scope of the PR but it shouldn't be too bad.

> I understand how you think but as a user it is weird to come to a editpage and not be able to save the changes. If the user only should be able to remove the activity a dialog like values is a much better solution.

I agree, the currently implementation is a weird experience for the user. My intentions with this PR was to add missing functionality to the registrator, not to specifically add the functionality to remove an activity. This was meant to be a draft, a mistake on my part, since the scope of the feature grew when i realised that the storage module did not have the required functionality. The remaining functionality will be added before the PR closes.

Neathan commented:
This should now also implement #72

Neathan commented:
You should now be able to both move and delete existing activities. I've implemented the changes suggested by Erik which made the remove function much simler and allowed for specific removal of a give activity. A move function was also added to the storage module which combines the 'remove-' and 'addInterval' functions to move a given activity.

malin-mallan reviewed (approved):
I havent read the code but i have run the code on both iOS and Androdi and its look really good now!
One thing that maye be a good feture is that the sliders stays at the correct number when edit. Now it´s only the number that is the number that was before and the sliders are set to 5.
